### ISSUE TITLE
`IndexError: list index out of range` when dropping inherited type with certain access policy present

### ISSUE THREAD

**Issue by jackfischer** [edgedb/edgedb#6404]
- EdgeDB Version: 3.5 and 4.0-rc.2
- EdgeDB CLI Version: 3.5.0+e1ad387
- OS Version: macos 13

Steps to Reproduce:

1. Migrate to base schema
2. Drop `Principal` and `extending Principal`
3. Create a migration, receive `list index out of range` error
```
$ edgedb migration create
did you alter object type 'default::User'? [y,n,l,c,b,s,q,?]
> y
Error applying statement: InternalServerError: IndexError: list index out of range
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
Rolling back last operation...
```

Base Schema:
```
using future nonrecursive_access_policies;

module default {

    global current_user: uuid;
    global current_user_object := (
        select User filter .id = global current_user
    );


    type Organization {
        access policy allow_select_users
            allow select
            using (global current_user_object.organization.id ?= .id);
    }

    abstract type Principal {   }
    type User  extending Principal {
    # type User {
        required link organization: Organization;
    }

}
```

**Comment by aljazerzen** [comment]
For future reference, creating migration from "base" schema to this schema fails:

```
using future nonrecursive_access_policies;

module default {

    global current_user: uuid;
    global current_user_object := (
        select User filter .id = global current_user
    );


    type Organization {
        access policy allow_select_users
            allow select
            using (global current_user_object.organization.id ?= .id);
    }

    type User {
        required link organization: Organization;
    }
}
```

**Comment by jackfischer** [comment]
@aljazerzen what's going on with the migration from base up to that schema? Is it the same underlying bug?

**Comment by aljazerzen** [comment]
During fixing of this bug I've encountered a different bug with how error messages are being composed (nothing major), which I fixed in #6425

I'm not sure this is the problem you are talking about, but it probably is.

---

To answer your other question from discord:

> ... I wanted to confirm this issue is not risky for correctness of access policies? We rely heavily on access policies in prod so if there's any risk we'd want to investigate it

No, this issue does not expose any access that should not be exposed according to access policies.

The effect of this bug is an Internal Server Error from *schema delta computation* for a specific case of defined schema, which means that the bug errors out before affecting SELECT or DML queries.
